text/0000-layout-packed-aligned.md

```diff
@@ -0,0 +1,176 @@
+- Feature Name: `layout_packed_aligned`
+- Start Date: 2024-10-24
+- RFC PR: [rust-lang/rfcs#3718](https://github.com/rust-lang/rfcs/pull/3718)
+- Rust Issue: [rust-lang/rust#100743](https://github.com/rust-lang/rust/issues/100743)
+
+# Summary
+[summary]: #summary
+
+This RFC makes it legal to have `#[repr(C)]` structs that are:
+- Both packed and aligned.
+- Packed, and transitively contains`#[repr(align)]` types.
+
+It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.
+It has the same behavior as `#[repr(C)]` except on `*-pc-windows-gnu` targets where it uses the msvc layout
+rules instead.
+
+# Motivation
+[motivation]: #motivation
+
+This RFC enables the following struct definitions:
+
+```rs
+#[repr(C, packed(2), align(4))]
+struct Foo { // Alignment = 4, Size = 8
+    a: u8,   // Offset = 0
+    b: u32,  // Offset = 2
+}
+```
+
+This is commonly needed when Rust is being used to interop with existing C and C++ code bases, which may contain
+unaligned types. For example in `clang` it is possible to create the following type definition, and there is
+currently no easy way to create a matching Rust type:
+
+```cpp
+struct  __attribute__((packed, aligned(4))) MyStruct {
+    uint8_t a;
+    uint32_t b;
+};
+```
+
+Currently `#[repr(packed(_))]` structs cannot transitively contain #[repr(align(_))] structs due to differing behavior between msvc and gcc/clang.
+However, in most cases, the user would expect `#[repr(C)]` to produce a struct layout matching the same type as defined by the current target. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## `#[repr(C)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target toolchain ABI.
+
+For example, given:
+```c
+#[repr(C, align(4))]
+struct Foo(u8);
+#[repr(C, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and 1 for everything else.
+
+
+## `#[repr(system)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target OS ABI.
+
+For example, given:
+```c
+#[repr(C, align(4))]
+struct Foo(u8);
+#[repr(C, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and `*-pc-windows-gnu`. It would be 1 for everything else.
```

Comment:
Are the structs in these examples supposed to be `#[repr(system, ...)]`

Reply:
Good catch, thanks!